src/chunk.js

```diff
@@ -0,0 +1,209 @@
+  }
+}
+
+class Chunks {
+  /**
+   * @param {!./service/ampdoc-impl.AmpDoc} ampDoc
+   */
+  constructor(ampDoc) {
+    /** @private @const */
+    this.ampDoc_ = ampDoc;
+    /** @private @const {!Window} */
+    this.win_ = ampDoc.win;
+    /** @private @const {!Array<function()>} */
+    this.tasks_ = [];
+    /** @private {?./service/viewer-impl.Viewer} */
+    this.viewer_ = null;
+    /** @private @const {function()} */
+    this.boundExecute_ = () => this.execute_();
+    /** @private @const {boolean} */
+    this.active_ = isExperimentOn(this.win_, 'chunked-amp');
+
+    if (!this.active_) {
+      return;
+    }
+    if (!this.win_.requestIdleCallback) {
+      this.win_.addEventListener('message', this.boundExecute_);
+    }
+    viewerPromiseForDoc(ampDoc).then(viewer => {
+      this.viewer_ = viewer;
+      viewer.onVisibilityChanged(() => {
+        if (viewer.isVisible()) {
+          this.execute_();
+        }
+      });
+      if (viewer.isVisible()) {
+        this.execute_();
+      }
+    });
+  }
+
+  /**
+   * Run fn as a "chunk". It'll run in a micro task when the doc is visible
+   * and otherwise run it after having yielded to the event queue once.
+   * @param {function()} fn
+   * @private
+   */
+  run_(fn) {
+    this.tasks_.push(fn);
+    this.schedule_();
+  }
+
+  /**
+   * Run a task.
+   * Schedule the next round if there are more tasks.
+   * @return {boolean} Whether anything was executed.
+   * @private
+   */
+  execute_() {
+    const t = this.tasks_.shift();
+    if (!t) {
+      return false;
+    }
+    const before = Date.now();
+    try {
+      t();
+    } catch (e) {
+      // We run early in init. All errors should show the doc.
+      makeBodyVisible(self.document);
+      throw e;
+    } finally {
+      if (this.tasks_.length) {
+        this.schedule_();
+      }
+      dev().fine('chunk', t.displayName || t.name,
+          'Duration', Date.now() - before);
+    }
+    return true;
+  }
+
+  /**
+   * Schedule running a task.
+   * @private
+   */
+  schedule_() {
+    if (!this.active_ || this.isVisible_()) {
+      resolved.then(this.boundExecute_);
+      return;
+    }
+    // If requestIdleCallback exists, schedule a task with it, but
+    // do not wait longer than one second.
+    if (this.win_.requestIdleCallback) {
+      this.win_.requestIdleCallback(this.boundExecute_, {
+        timeout: 1000,
+      });
+      return;
+    }
+    // The message doesn't actually matter.
+    this.win_.postMessage/*OK*/('amp-macro-task', '*');
+  }
+
+  /**
+   * @return {boolean}
+   * @private
+   */
+  isVisible_() {
+    // Ask the viewer first.
+    if (this.viewer_) {
+      return this.viewer_.isVisible();
+    }
+    // There is no viewer yet. Lets try to guess whether we are visible.
+    if (this.win_.document.hidden) {
+      return false;
+    }
+    // Viewers send a URL param if we are not visible.
+    return !(/visibilityState=hidden/.test(this.win_.location.hash));
```

Comment:
Not sure if I already commented on this, but there are many other values that might be of important here: `prerender`, `paused`, `inactive`. Probably the most important is `prerender` since this is the state most of docs start in. Are you counting on the fact that it will be rare that `viewer_` is not available yet to respond?